### ISSUE TITLE
addon-backgrounds: global context not initially updated with color

### ISSUE THREAD

**Issue by nickofthyme** [storybookjs/storybook#15632]
**Describe the bug**
When using [`@storybook/addons-backgrounds`](https://storybook.js.org/docs/react/essentials/backgrounds) in a react project with a custom `Decorator`, the initial/default background value is not passed to the `context.globals`.

I have a project performing text contrast logic where I need the initial value to pass to javascript.

**To Reproduce**
1. Pull down and run this repro [here](url).
2. Go to the primary button story.
3. Notice the `console.log` of the `context.globals` shows and empty object.
4. Click on a different background color, notice the `context.globals` now includes the selected value.

**System**

```
Environment Info:

  System:
    OS: macOS 11.4
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  Binaries:
    Node: 14.17.3 - ~/.nvm/versions/node/v14.17.3/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.13 - ~/.nvm/versions/node/v14.17.3/bin/npm
  Browsers:
    Chrome: 91.0.4472.114
    Firefox: 90.0.1
    Safari: 14.1.1
  npmPackages:
    @storybook/addon-backgrounds: ^6.3.4 => 6.3.4
    @storybook/react: ^6.3.4 => 6.3.4
```

**Additional context**
Could be related to #14846

**Comment by nickofthyme** [comment]
After some digging, it looks like the so-called `_initialGlobals` are filtered by `allowedGlobals` which is determined from `parameters.globals`???? I don't see this documented anywhere in the docs, see code below.

https://github.com/storybookjs/storybook/blob/209b6c4cb3aa9dbc73a7c971fcf71e95a51a7065/lib/client-api/src/story_store.ts#L216-L228

@ghengeveld could you shed some light on this per you changes in #15056?

Basically, adding `globals.backgrounds` like below enables the query param global value.

```diff
export const parameters = {
+  globals: {
+    backgrounds: {}
+  },
  backgrounds: {
```

---

Still doesn't solve the default case where no global param is present.

**Comment by shilman** [comment]
this looks like a bug with `addon-backgrounds` or with core, depending on how you see it cc @yannbf @tmeasday @ghengeveld 

also cc @stevensacks since you're using the same pattern in your i18n work

AFAICT what's going on here is that the globals code in core assumes that the globals are declared statically, either by the end user or by addons. what that would look like:

```js
// .storybook/preview.js
export const globalTypes = { ... };
export const globals = { ... };
```

If you don't declare a `globalType`, it will infer the type from the data in `globals`. So the final results union these two.

`addon-backgrounds` uses a different pattern where the user declares global values as parameters, and then the addon calls `updateGlobals` on story render. I'm not a fan of this pattern, but it's what we have for now.

Given this pattern, we could fix the problem:
1. The backgrounds-addon could declare the globalType that it's about to populate on its first render.
2. The core code could be less strict about what it accepts.

I think the first fix is closer to the design intent of globals. WDYT?

**Comment by tmeasday** [comment]
I think I agree @shilman

**Comment by nickofthyme** [comment]
> `addon-backgrounds` uses a different pattern where the user declares global values as parameters, and then the addon calls `updateGlobals` on story render. I'm not a fan of this pattern, but it's what we have for now.

@shilman based on the inability to statically define the initial background color are you okay with the changes in #15640?

**Comment by bytrangle** [comment]
@nickofthyme Maybe I don't understand the issue very well, but in [your demo](https://github.com/nickofthyme/repro-sb-15632), you tried to get the value of `context.globals` in a decorator function. But you didn't declare a `globalTypes` in `.storybook/preview.js`. So `context.globals` is empty on initial rendering. Isn't this expected?

**Comment by nickofthyme** [comment]
@bytrangle the issue is that it does not capture the initial default value in `preview.js` [here](https://github.com/nickofthyme/repro-sb-15632/blob/86eea3d63e2d8ddb2bb309044308f0c8b54c4828/.storybook/preview.js#L5), does not get passed to the decorator.

The whole `globalTypes` vs `globals` is really confusing when to use one or the other or both. As `@shilman` said

> If you don't declare a `globalType`, it will infer the type from the data in `globals`. So the final results union these two.

Is this not true? If it is I'd think the demo would work as is.

---

All in all how do you make the [demo](https://github.com/nickofthyme/repro-sb-15632) work so that it passes the initial default value to the decorators?

**Comment by ndelangen** [comment]
Is this still an issue with 7.0 beta?

**Comment by lmcgartland** [comment]
+1 I'm still seeing this issue

**Comment by Sidnioulz** [comment]
Folks, could you please try again with Storybook 9? This could be a duplicate of https://github.com/storybookjs/storybook/issues/28592.

**Comment by bytrangle** [comment]
Is this issue accidentally solved?

**Comment by ndelangen** [comment]
I closed it because the default value for the global is now defined AS a global, so it will be there, no need to be taken from parameters, since that's no longer where it's defined.

**Comment by jack-bliss** [comment]
is this just going to be permanently broken for v7?

**Comment by Sidnioulz** [comment]
> is this just going to be permanently broken for v7?

Storybook 7 no longer receives feature / non-critical bug fix updates. Please update to a more recent version of Storybook.
